session.go

```diff
@@ -247,16 +248,19 @@ func (sess *session) X11() (X11, bool) {
 }
 
 func (sess *session) Signals(c chan<- Signal) {
-	sess.Lock()
-	defer sess.Unlock()
+	sess.sigMu.Lock()
 	sess.sigCh = c
-	if len(sess.sigBuf) > 0 {
-		go func() {
-			for _, sig := range sess.sigBuf {
-				sess.sigCh <- sig
-			}
-		}()
+	if len(sess.sigBuf) == 0 || sess.sigCh == nil {
+		sess.sigMu.Unlock()
+		return
 	}
+	go func() {
+		defer sess.sigMu.Unlock()
```

Comment:
this is unorthodox enough to warrant a comment; it looked like a bug to me at first.

Reply:
I tried to captured the motivation in https://github.com/coder/ssh/pull/5/commits/45772fad1b8d0f14b2ce7b5cf7542f77dbbca649 now, thoughts?

Reply:
what I meant is the fact that you're calling

```
defer sess.sigMu.Unlock()
```

without the corresponding `Lock()` in the same function looks very strange to me!  We often talk about a function "holding" the lock while it runs, but here you "transfer" the lock from the Signals() function to the child goroutine so the parent can return.

Reply:
Yes, I understood that's what you meant. Did you feel the motivation for it was lacking?

Reply:
As an aside, this is explicitly documented in https://pkg.go.dev/sync#Mutex.Unlock

> A locked Mutex is not associated with a particular goroutine. It is allowed for one goroutine to lock a Mutex and then arrange for another goroutine to unlock it.

Not that I'm making any claims about this being standard in any way... I was simply trying to retrofit an edge-cases covered fix into the existing code without too many changes.

Interestingly, there's been a proposal to disallow this, but it didn't gain traction: https://github.com/golang/go/issues/9201

Reply:
It was actually just about the mechanics --- like

```
// here we're relying on the fact that the mutex was locked in the outer `Signal()`
// function, so this goroutine/function just has to unlock the mutex when it is done
```